### PR TITLE
Adds CBMC proofs for aws_cryptosdk_priv_hdr_parse_{auth_tag,iv,alg_suite_data}

### PR DESCRIPTION
**Pull request by tegansb**
Adds harness, Makefile and yaml for aws_cryptosdk_priv_hdr_parse_auth_tag, aws_cryptosdk_priv_hdr_parse_iv, and aws_cryptosdk_priv_hdr_parse_alg_suite_data.

Uses ensures, is bounded and validity functions from #632.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: source/header.c [hunk 1]
```diff
@@ -333,6 +333,9 @@ int aws_cryptosdk_priv_hdr_parse_frame_len(
 
 int aws_cryptosdk_priv_hdr_parse_alg_suite_data(
     struct aws_cryptosdk_hdr *hdr, const struct aws_cryptosdk_alg_properties *alg_props, struct aws_byte_cursor *cur) {
+    AWS_PRECONDITION(aws_cryptosdk_hdr_is_valid(hdr));
+    AWS_PRECONDITION(&hdr->alloc != NULL);
```
Comment (feliperodri) [inline source/header.c:337]: You don't need `&hdr->alloc`, right? Why not just `hdr->alloc != NULL`?

File: source/header.c [hunk 2]
```diff
@@ -342,12 +345,18 @@ int aws_cryptosdk_priv_hdr_parse_alg_suite_data(
 }
 
 int aws_cryptosdk_priv_hdr_parse_iv(struct aws_cryptosdk_hdr *hdr, uint8_t iv_len, struct aws_byte_cursor *cur) {
+    AWS_PRECONDITION(aws_cryptosdk_hdr_is_valid(hdr));
+    AWS_PRECONDITION(&hdr->alloc != NULL);
```
Comment (feliperodri) [inline source/header.c:349]: You don't need `&hdr->alloc`, right? Why not just `hdr->alloc != NULL`?

File: verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_alg_suite_data/aws_cryptosdk_priv_hdr_parse_alg_suite_data_harness.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+void aws_cryptosdk_priv_hdr_parse_alg_suite_data_harness() {
+    /* Nondet Input */
+    struct aws_cryptosdk_hdr *hdr   = hdr_setup(MAX_TABLE_SIZE, MAX_EDK_LIST_ITEMS, MAX_BUFFER_SIZE);
+    struct aws_byte_cursor *pcursor = malloc(sizeof(*pcursor));
+    enum aws_cryptosdk_alg_props alg_id;
+    struct aws_cryptosdk_alg_properties *alg_props = aws_cryptosdk_alg_props(alg_id);
+
+    /* Assumptions */
+    __CPROVER_assume(&hdr->alloc != NULL);
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_alg_suite_data/aws_cryptosdk_priv_hdr_parse_alg_suite_data_harness.c:31]: You don't need `&hdr->alloc`, right? Why not just `hdr->alloc != NULL`?

File: verification/cbmc/sources/make_common_data_structures.c [hunk 4]
```diff
@@ -178,6 +178,8 @@ struct aws_cryptosdk_hdr *ensure_nondet_hdr_has_allocated_members(const size_t m
     struct aws_cryptosdk_hdr *hdr = malloc(sizeof(*hdr));
     if (hdr != NULL) {
         hdr->alloc = nondet_bool() ? NULL : can_fail_allocator();
+        enum aws_cryptosdk_alg_props alg_id;
+        hdr->alg_id = alg_id;
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:182]: Why do you need that here? If `alg_id` is a enum, than CBMC will assign a non-det value for it by default.

File: verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_iv/aws_cryptosdk_priv_hdr_parse_iv_harness.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#include <proof_helpers/utils.h>
+
+void aws_cryptosdk_priv_hdr_parse_iv_harness() {
+    /* Nondet Input */
+    struct aws_cryptosdk_hdr *hdr   = hdr_setup(MAX_TABLE_SIZE, MAX_EDK_LIST_ITEMS, MAX_BUFFER_SIZE);
+    struct aws_byte_cursor *pcursor = malloc(sizeof(*pcursor));
+    uint8_t iv_len;
+
+    /* Assumptions */
+    __CPROVER_assume(&hdr->alloc != NULL);
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_iv/aws_cryptosdk_priv_hdr_parse_iv_harness.c:30]: You don't need `&hdr->alloc`, right? Why not just `hdr->alloc != NULL`?

**Review by feliperodri — CHANGES_REQUESTED** [review]
I have a few questions.

File: source/header.c (hunk 1 above)
Comment (alex-chew) [reply source/header.c:337]: It seems to me that the `alloc` precondition isn't necessary at all, whether @tegansb meant `&hdr->alloc` or `hdr->alloc` - the `aws_cryptosdk_hdr_is_valid(hdr)` on the previous line implies that

1. `hdr != NULL`, which in turn implies `&hdr->alloc != NULL`
2. `aws_allocator_is_valid(hdr->alloc)`, which in turn implies `hdr->alloc != NULL` according to https://github.com/awslabs/aws-c-common/blob/d0b5592e6b1efbd5fce4ee2cc8fa4aecf297e54c/source/allocator.c#L28

**Review by alex-chew — COMMENTED** [review]
Mostly LGTM, though I'm curious about the `&hdr->alloc != NULL` preconditions that Felipe pointed out. I think that the `aws_cryptosdk_hdr_is_valid(hdr)` precondition is sufficient, and it's present in both `header.c` (in the line above) and in the proof harnesses (via `hdr_setup`).

File: source/header.c (hunk 1 above)
Comment (tegansb) [reply source/header.c]: You're right! I think I added this assumption before the aws_allocator_is_valid one and didn't realize I could remove it once the validity assumption was added. I've removed it now from both the harnesses and the source code.

File: source/header.c (hunk 2 above)
Comment (tegansb) [reply source/header.c]: See above!

File: verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_alg_suite_data/aws_cryptosdk_priv_hdr_parse_alg_suite_data_harness.c (hunk 3 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_alg_suite_data/aws_cryptosdk_priv_hdr_parse_alg_suite_data_harness.c]: See above!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 4 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c]: Gotcha, I've removed this now. You're right: it isn't needed.

File: verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_iv/aws_cryptosdk_priv_hdr_parse_iv_harness.c (hunk 5 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_priv_hdr_parse_iv/aws_cryptosdk_priv_hdr_parse_iv_harness.c]: See above.

File: source/header.c [hunk 6]
```diff
@@ -354,6 +354,8 @@ int aws_cryptosdk_priv_hdr_parse_frame_len(
 
 int aws_cryptosdk_priv_hdr_parse_alg_suite_data(
     struct aws_cryptosdk_hdr *hdr, const struct aws_cryptosdk_alg_properties *alg_props, struct aws_byte_cursor *cur) {
+    AWS_PRECONDITION(aws_cryptosdk_hdr_is_valid(hdr));
```
Comment (adpaco-aws) [inline source/header.c:357]: Don't you needs `alg_props` to be valid here? I see the assumption in the proof but not here.

**Review by adpaco-aws — CHANGES_REQUESTED** [review]
Just a minor issue, otherwise LGTM.

File: source/header.c (hunk 6 above)
Comment (tegansb) [reply source/header.c:357]: Added!

**Review by adpaco-aws — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]
LGTM.

**Review by alex-chew — APPROVED** [review]
LGTM

**Pull request merged**